### PR TITLE
upload wechat notification plugin

### PR DESCRIPTION
**Pull request by Martindlut**
### Status 

- new pack, complete

### Description

Add new action which sends messages to Wechat Enterprise Account

### Checklist (tick everything that applies)

- [x] Metadata: pack.yaml, icon, structure (required for new packs)
- [x] Version bump (required for changed packs)
- [x] Code linting (required, can be done after the PR checks)
- [ ] [Tests](https://docs.stackstorm.com/development/pack_testing.html) (not required but really recommended)

### CODE REVIEW

**Comment by Kami** [comment]
Thanks for the contribution!

We will have a look shortly.

In the mean time please make sure the build is green (right now it looks like some lint checks are failing).

File: packs/wechat/actions/lib/wechatutil/getappinfo.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+from lib.wechatutil.wechatapp import WechatApplication
+apiUrl = 'https://qyapi.weixin.qq.com/cgi-bin/agent/get'
+
```
Comment (LindsayHill) [inline packs/wechat/actions/lib/wechatutil/getappinfo.py]: Please add another blank line here

File: packs/wechat/actions/lib/wechatutil/getappinfo.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+def getappinfo(id, factory):
+    payload = {"agentid": id}
+    response = factory.get(apiUrl, payload)
+    if response['errcode'] != '0':
+        factory.refreshtoken();
```
Comment (LindsayHill) [inline packs/wechat/actions/lib/wechatutil/getappinfo.py]: Don't think the ';' is needed here?

File: packs/wechat/actions/lib/wechatutil/getappinfo.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+        response = factory.get(apiUrl, payload)
+    if response['errcode'] == '0':
+        return 0, WechatApplication(response), None
+    else:
+        return int(response['errcode']), None, response['errmsg']
+
```
Comment (LindsayHill) [inline packs/wechat/actions/lib/wechatutil/getappinfo.py]: Blank line needs to be removed

File: packs/wechat/actions/lib/wechatutil/gettoken.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+
+
+def gettoken(corpid, secret, session):
+    header = {'content-type': 'application/json; charset=utf8'}
+    payload = {"corpid": corpid, "corpsecret": secret}
+    response = session.get(apiUrl, params=payload, headers=header).json()
+    if response['access_token'] is not None:
+        return 0, response['access_token'], None
+    else:
+        return response['errcode'], None, response['errmsg']
```
Comment (LindsayHill) [inline packs/wechat/actions/lib/wechatutil/gettoken.py]: needs a newline at the end here (nb not a blank line, just a newline at end of line)

File: packs/wechat/actions/lib/wechatutil/message.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    def setcontent(self, *articles):
+        self.args[self.msgtype] = {"articles": articles}
+
+
+class MpNewsMessage(Message):
+    msgtype = 'mpnews'
+
+    def setcontent(self, *articles):
+        self.args[self.msgtype] ={"articles": articles}
+
```
Comment (LindsayHill) [inline packs/wechat/actions/lib/wechatutil/message.py]: needs an extra blank line in here

File: packs/wechat/actions/lib/wechatutil/message.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    def setcontent(self, *articles):
+        self.args[self.msgtype] = {"articles": articles}
+
+
+class MpNewsMessage(Message):
+    msgtype = 'mpnews'
+
+    def setcontent(self, *articles):
+        self.args[self.msgtype] ={"articles": articles}
```
Comment (LindsayHill) [inline packs/wechat/actions/lib/wechatutil/message.py]: needs space after '='

File: packs/wechat/actions/lib/wechatutil/message.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+class FileMessage(ResourceMessage):
+    msgtype = 'file'
+
+
+class VideoMessage(Message):
+    msgtype = 'video'
+
+    def setcontent(self, mediaid, title, description):
+        self.args[self.msgtype]={'media_id': mediaid, 'title': title, 'description': description}
```
Comment (LindsayHill) [inline packs/wechat/actions/lib/wechatutil/message.py]: needs spaces around =

File: packs/wechat/actions/lib/wechatutil/message.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+class TextMessage(Message):
+    msgtype = 'text'
+
+    def setcontent(self,msg):
+        self.args[self.msgtype] = {"content": msg}
+
+class ResourceMessage(Message):
+    def setcontent(self,mediaid):
+        self.args[self.msgtype] ={"media_id": mediaid}
```
Comment (LindsayHill) [inline packs/wechat/actions/lib/wechatutil/message.py]: needs space after =

File: packs/wechat/actions/lib/wechatutil/message.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+
+class TextMessage(Message):
+    msgtype = 'text'
+
+    def setcontent(self,msg):
+        self.args[self.msgtype] = {"content": msg}
+
+class ResourceMessage(Message):
+    def setcontent(self,mediaid):
```
Comment (LindsayHill) [inline packs/wechat/actions/lib/wechatutil/message.py]: needs space after ','

File: packs/wechat/actions/lib/wechatutil/message.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        self.args['agentid'] = agentid
+
+    def setmsgsafe(self):
+        self.args['safe'] = 1
+
+
+class TextMessage(Message):
+    msgtype = 'text'
+
+    def setcontent(self,msg):
```
Comment (LindsayHill) [inline packs/wechat/actions/lib/wechatutil/message.py]: needs space after ','

File: packs/wechat/actions/lib/wechatutil/message.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        self.args[self.msgtype] = {"content": msg}
+
```
Comment (LindsayHill) [inline packs/wechat/actions/lib/wechatutil/message.py]: needs one more blank line in here

File: packs/wechat/actions/lib/wechatutil/sendmessages.py [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+apiUrl = 'https://qyapi.weixin.qq.com/cgi-bin/message/send'
+
+
+def send_msg(message, factory):
+    response = factory.post(apiUrl, data=json.dumps(message.args))
+    if response['errcode'] != 0:
+        factory.refreshtoken()
+        response = factory.post(apiUrl, data=json.dumps(message.args))
+    return response['errcode'], response['errmsg']
+
```
Comment (LindsayHill) [inline packs/wechat/actions/lib/wechatutil/sendmessages.py]: Could you remove the blank lines at the end of this file?

File: packs/wechat/actions/lib/wechatutil/wechatapp.py [hunk 13]
```diff
@@ -0,0 +1,27 @@
+from lib.wechatutil.appinfo import AppInfo
+class WechatApplication:
```
Comment (LindsayHill) [inline packs/wechat/actions/lib/wechatutil/wechatapp.py]: Could you add two blank lines in here?

File: packs/wechat/actions/lib/wechatutil/wechatapp.py [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    def __init__(self,**kwargs):
```
Comment (LindsayHill) [inline packs/wechat/actions/lib/wechatutil/wechatapp.py]: and a space here after "self,"

File: packs/wechat/actions/lib/wechatutil/wechatapp.py [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+        return self.appinfo.getagentid()
+
+    def getallow_userinfos(self):
+        return self.getallow_userinfos()
+
+    def getallow_partys(self):
+        return self.allow_partys
+
+    def getallow_tags(self):
+        return self.allow_tags
```
Comment (LindsayHill) [inline packs/wechat/actions/lib/wechatutil/wechatapp.py]: And a newline character at the end of the file (not a blank line)

File: packs/wechat/actions/lib/wechatutil/wechatclient.py [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+
+    def sendTextMessage(self, agentid, content, **receivers):
+        msg = TextMessage()
+        msg.setagentid(agentid)
+        msg.setcontent(content)
+        msg.setreceiver(**receivers)
+        rc, msg = send_msg(msg, self.connfactory)
+        if rc != 0:
+            raise ValueError(msg)
+
```
Comment (LindsayHill) [inline packs/wechat/actions/lib/wechatutil/wechatclient.py]: Can you remove these blank lines?

File: packs/wechat/actions/lib/wechatutil/wechatclient.py [hunk 17]
```diff
@@ -0,0 +1,24 @@
+from lib.wechatutil.sendmessages import send_msg
+from lib.wechatutil.wechatrequest import WechatRequest
+from lib.wechatutil.message import *
```
Comment (LindsayHill) [inline packs/wechat/actions/lib/wechatutil/wechatclient.py]: The "import *" part seems to be causing some problems with tests. Can we make this more specific?

File: packs/wechat/actions/lib/wechatutil/wechatrequest.py [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+        self.refreshtoken()
+
+    def get(self, url, customparams, customheader=None):
+        header = self.defaultheader
+        params = {"access_token": self.token}
+        params.update(customparams)
+        if customheader is not None:
+            header.update(customheader)
+        response = self.session.get(url, params=params, headers=header)
+        return response.json();
```
Comment (LindsayHill) [inline packs/wechat/actions/lib/wechatutil/wechatrequest.py]: Looks like the ';' is not needed

File: packs/wechat/actions/lib/wechatutil/wechatrequest.py [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+
+    def post(self, url, data, customheader=None):
+        header = self.defaultheader
+        if customheader is not None:
+            header.update(customheader)
+        response = self.session.post(url+"?access_token="+self.token, data=data, headers=header)
```
Comment (LindsayHill) [inline packs/wechat/actions/lib/wechatutil/wechatrequest.py]: Needs spaces around `+`

File: packs/wechat/actions/wechatnotifier.py [hunk 20]
```diff
@@ -0,0 +1,20 @@
+from st2actions.runners.pythonrunner import Action
+from lib.wechatutil.wechatclient import WechatClient
+class WechatNotifierAction(Action):
```
Comment (LindsayHill) [inline packs/wechat/actions/wechatnotifier.py]: two newlines needed here

File: packs/wechat/actions/wechatnotifier.py [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+    def run(self, corpID, corpSecret, agentid, message, touser=None, togroup=None, totag=None):
+        receivers=dict()
```
Comment (LindsayHill) [inline packs/wechat/actions/wechatnotifier.py]: space needed around `=`

File: packs/wechat/actions/wechatnotifier.py [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+        if touser:
+            receivers={"touser": touser}
```
Comment (LindsayHill) [inline packs/wechat/actions/wechatnotifier.py]: and same here - space needed

File: packs/wechat/actions/wechatnotifier.py [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+            receivers.update({"toparty": togroup})
+
+        if totag:
+            receivers.update({"totag": totag})
+
+        if len(receivers) == 0:
+            raise KeyError('touser, togroup, totag can not be empty at the same time')
+
+        wechatClient=WechatClient(corpID, corpSecret)
+        wechatClient.sendTextMessage(agentid, message,**receivers)
```
Comment (LindsayHill) [inline packs/wechat/actions/wechatnotifier.py]: Spaces needed around `=` and after `,`

**Comment by LindsayHill** [comment]
@Martindlut Could you please make the changes I've suggested? That should get most of the linting problems sorted out. Or would you like me to take over the PR? Or I can submit a PR to your branch?

**Comment by LindsayHill** [comment]
Since there has been no recent activity on this PR, I'm taking it over in #554

**Pull request closed without merging**